### DISCUSSION TITLE
How to pass output to caller argument. (IE how to mock Read trait)

### DISCUSSION THREAD

**Discussion by TomzBench** [asomers/mockall#303] [Questions] [answered]
Hello, I have a function i am trying to mock that is similar to the Read trait. 

```
   fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
 ```

My mock needs to populate the caller supplied buffer. I see I can return arbitrary value, but how can I manipulate the caller arguments?

**Comment by asomers** [comment, marked as answer]
Just modify it as a side effect with .returning.  Like this:
```rust
foo.expect_read()
    .returning(|mut buf| {
        buf[0] = 42;
        Ok(1)
    })
```
